lib/iris/analysis/_regrid.py

```diff
@@ -287,83 +288,105 @@ def _regrid_indices(cells, depth, points):
             # Calculate a new 'sum_weights' to allow for missing source points.
             # N.B. it is more efficient to use the original once-calculated
             # sparse matrix, but in this case we can't.
             # Hopefully, this post-multiplying by the validities is less costly
             # than repeating the whole sparse calculation.
-            valid_src_cells = ~mask.flat[:]
-            src_cell_validity_factors = sparse_diags(
-                np.array(valid_src_cells, dtype=int), 0
-            )
-            valid_weights = sparse_matrix * src_cell_validity_factors
-            sum_weights = valid_weights.sum(axis=1).getA()
-            # Work out where output cells are missing all contributions.
-            # This allows for where 'rows' contains output cells that have no
-            # data because of missing input points.
-            zero_sums = sum_weights == 0.0
-            # Make sure we can still divide by sum_weights[rows].
-            sum_weights[zero_sums] = 1.0
+            valid_src_cells = ~mask.reshape(-1, grid_size)
+            sum_weights = valid_src_cells @ sparse_matrix.T
+        data = r_data
+    # Work out where output cells are missing all contributions.
+    # This allows for where 'rows' contains output cells that have no
+    # data because of missing input points.
+    zero_sums = sum_weights == 0.0
+    # Make sure we can still divide by sum_weights[rows].
+    sum_weights[zero_sums] = 1.0
 
     # Calculate sum in each target cell, over contributions from each source
     # cell.
-    numerator = sparse_matrix * data.reshape(-1, 1)
-
-    # Create a template for the weighted mean result.
-    weighted_mean = ma.masked_all(numerator.shape, dtype=numerator.dtype)
-
-    # Calculate final results in all relevant places.
-    weighted_mean[rows] = numerator[rows] / sum_weights[rows]
-    if is_masked:
-        # Ensure masked points where relevant source cells were all missing.
-        if np.any(zero_sums):
-            # Make masked if it wasn't.
-            weighted_mean = np.ma.asarray(weighted_mean)
-            # Mask where contributing sums were zero.
-            weighted_mean[zero_sums] = np.ma.masked
-
-    # Construct the final regridded weighted mean cube.
+    numerator = data.reshape(-1, grid_size) @ sparse_matrix.T
+
+    weighted_mean = numerator / sum_weights
+    # Ensure masked points where relevant source cells were all missing.
+    weighted_mean = ma.asarray(weighted_mean)
+    if np.any(zero_sums):
+        # Mask where contributing sums were zero.
+        weighted_mean[zero_sums] = ma.masked
+
+    new_data_shape = list(data_shape)
+    for dim, length in zip(inds, grid_cube.shape):
+        new_data_shape[dim] = length
+    if len(dims) == 1:
+        new_data_shape.append(grid_cube.shape[1])
+        dims = (dims[0], dims[0] + 1)
+    if len(dims) > 2:
+        new_data_shape = new_data_shape[: 2 - len(dims)]
+        dims = dims[:2]
+    inds = [-2, -1]
```

Comment:
Was this hardcoding of `inds` leftover from some debugging, perhaps?

Reply:
Yeah, I think this was rewritten a few times and I suppose it could be made clearer.